### PR TITLE
[v6r19] Add submit options

### PR DESCRIPTION
**Pull request by schr476**
Modified the GlobusComputingElement.py to include more globus-job-submit options.
Required to properly use HPC systems.

### CODE REVIEW

File: Resources/Computing/GlobusComputingElement.py [hunk 1]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
     for _i in xrange(numberOfJobs):
       diracStamp = makeGuid()[:8]
       queueName = '%s/%s' % ( self.ceName, self.queue )
-      cmd = ['globus-job-submit', queueName, "-s", executableFile ]
+      #cmd = ['globus-job-submit', queueName, "-s", executableFile ]
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Please drop the commented line

File: Resources/Computing/GlobusComputingElement.py [hunk 2]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
+      cmd = ['globus-job-submit', queueName]
+      ## Add extra options
+      queueCount = self.ceParameters['QueueCount']
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Use `get` so there is no crash if the key is not present

File: Resources/Computing/GlobusComputingElement.py [hunk 3]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
+      queueHostCount = self.ceParameters['QueueHostCount']
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Use `get`

File: Resources/Computing/GlobusComputingElement.py [hunk 4]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
+      queuePar = self.ceParameters['QueueParameter']
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Use `get`

File: Resources/Computing/GlobusComputingElement.py [hunk 5]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
+      queueMaxTime = self.ceParameters['maxCPUTime']
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Use `get`

File: Resources/Computing/GlobusComputingElement.py [hunk 6]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
+      #queueOut = self.ceParameters['outLog']
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Drop the commented lines

File: Resources/Computing/GlobusComputingElement.py [hunk 7]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
+
+      if queueCount:
+        cmd += ['-count',queueCount]
+      if queueHostCount:
+        cmd += ['-host-count',queueHostCount] 
+      if queuePar:
+        cmd += ['-q',queuePar]
+      if queueMaxTime:
+        cmd += ['-maxtime', queueMaxTime]
+      #if queueOut:
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Drop the commented lines

File: Resources/Computing/GlobusComputingElement.py [hunk 8]
```diff
@@ -67,7 +67,35 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
+      #  cmd += ['-stdout', queueOut]
+      #if queueErr:
+      #  cmd += ['-stderr', queueErr]
+      cmd += ["-s", executableFile ]
+      self.log.verbose(cmd)
+      print(cmd)
```
Comment (andresailer) [inline Resources/Computing/GlobusComputingElement.py]: Drop the `print`

**Comment by fstagni** [comment]
I think v6r18 can be targeted here, it looks like a pretty easy/safe patch.

**Comment by andresailer** [comment]
Sure, after the proposed changes have been implemented and reviewed again :)

**Comment by atsareg** [comment]
I suggest to review this changes with the following respect: introduce "extraSubmitOptions" configuration option for ARC queues and use it as a catch all setting which is appended to the job submit command. In this case this will accommodate the introduced cases and any future case as well. Similar pattern can be used for some other ComputingElements.

**Comment by andresailer** [comment]
Why ARC? This is for Globus?

**Comment by atsareg** [comment]
Ah, sorry, Globus. But the same pattern can be applied to ARC as well :).

**Comment by atsareg** [comment]
In ARC we have, for example, XRSLExtraString "catch all" option

**Comment by fstagni** [comment]
I moved the target to v6r18, please @schr476 implement the suggested changes.

**Comment by schr476** [comment]
Sorry for the delay.
I've update most of the changes and testing this week.

**Comment by schr476** [comment]
Why does GlobusComputingElement require PilotAgentsDB?
Wouldn't be better to not rely on PilotAgentsDB and create a GlobusComputeDB, agreed?

**Comment by andresailer** [comment]
What would GlobusComputeDB do?

**Comment by atsareg** [comment]
Globus uses PilotAgentsDB because there is no ( I guess ) simple method to interrogate the status of a Globus CE with respect to a given VO. However, I think that this can be suppressed now that SiteDirector can be instructed not to get the status from the CE itself but rather from the PilotAgentsDB. The getCEStatus() can just return zero values but the Globus CE can be configured with QueryCEFlag False. This will force the SiteDirector to rely on the PilotAgentsDB only and the GlobusComputingElement can drop the PilotAgentsDB dependency.

**Comment by schr476** [comment]
If you drop the PilotAgentsDB with a flag then how does DIRAC keep track of the PilotJobReference info? Cached?

**Comment by fstagni** [comment]
Up!

**Comment by schr476** [comment]
Simple update to allow ExtraString to be added in the cfg.
Tested using NERSC Cori system (U.S. HPC machine).

File: Resources/Computing/GlobusComputingElement.py [hunk 9]
```diff
@@ -12,7 +12,7 @@
 
 """
 
-__RCSID__ = "$Id$"
+__RCSID__ = "338f3a4 (2015-10-06 13:34:21 +0200) Andre Sailer <andre.philippe.sailer@cern.ch>"
```
Comment (fstagni) [inline Resources/Computing/GlobusComputingElement.py]: No

File: Resources/Computing/GlobusComputingElement.py [hunk 10]
```diff
@@ -67,35 +69,9 @@ def submitJob( self, executableFile, proxy, numberOfJobs = 1 ):
...
-        cmd += ['-q',queuePar]
-      if queueMaxTime:
-        cmd += ['-maxtime', queueMaxTime]
-      #if queueOut:
-      #  cmd += ['-stdout', queueOut]
-      #if queueErr:
-      #  cmd += ['-stderr', queueErr]
-      cmd += ["-s", executableFile ]
+      
+      cmd = ['globus-job-submit', queueName, " ", self.extraSubmitString, "-s", executableFile ]
```
Comment (fstagni) [inline Resources/Computing/GlobusComputingElement.py]: Why dropping all the previous options?

Comment (schr476) [reply Resources/Computing/GlobusComputingElement.py]: Easier to add everything in one inclusive cfg.

Comment (andresailer) [reply Resources/Computing/GlobusComputingElement.py]: Why the empty `" ",` entry in the cmd list?

**Comment by fstagni** [comment]
\> 1 year old, no updates, no answer. Closing (comments still possible)

**Pull request closed without merging**